core/Segment.php

```diff
@@ -110,6 +110,35 @@ public function __construct($segmentCondition, $idSites)
         }
     }
 
+    private function getAvailableSegments()
+    {
+        // segment metadata
+        if (empty($this->availableSegments)) {
+            $this->availableSegments = API::getInstance()->getSegmentsMetadata($this->idSites, $_hideImplementationData = false);
```

Comment:
You could include the idsites in the cache key. I think it's possible for multiple segments to be created with the same idsites but different definition in one request.